File: 2.GammaCategorization/CreationalPatterns/5.Singleton/Aula07_SingletonExcercise/Aula_08_SinlgetonExercise2/C#/Main.cs
using System;

namespace CentralDeAlertasNamespace
{
    class Program
    {
        static void Main(string[] args)
        {
            // Criando objetos dos órgãos
            POLICIA POLICIA = new POLICIA();
            BOMBEIRO BOMBEIRO = new BOMBEIRO();
            SAMU SAMU = new SAMU();

            // Comprovando que a CentralDeAlertas é a mesma instância através do GetHashCode()
            Console.WriteLine("=== COMPROVAÇÃO DE INSTÂNCIA ÚNICA (HASHCODE) ===");
            Console.WriteLine($"Hash via Polícia:   {CentralDeAlertas.Instancia.GetHashCode()}");
            Console.WriteLine($"Hash via BOMBEIRO: {CentralDeAlertas.Instancia.GetHashCode()}");
            Console.WriteLine($"Hash via SAMU:      {CentralDeAlertas.Instancia.GetHashCode()}");
            Console.WriteLine("===============================================\n");

            // Cada órgão envia um alerta
            POLICIA.EnviarAlerta("Ocorrência de roubo na Zona Sul.");
            BOMBEIRO.EnviarAlerta("Princípio de incêndio em edificação comercial.");
            SAMU.EnviarAlerta("Acidente de trânsito grave na Rodovia Principal.");

            // Listando a partir de órgãos diferentes para provar o compartilhamento do histórico
            Console.WriteLine("--- Histórico visualizado pela POLÍCIA ---");
            foreach (var alerta in POLICIA.ListarAlertas())
            {
                Console.WriteLine(alerta);
            }

            Console.WriteLine("\n--- Histórico visualizado pelo SAMU ---");
            foreach (var alerta in SAMU.ListarAlertas())
            {
                Console.WriteLine(alerta);
            }
            
            Console.ReadKey();
        }
    }
}

File: 2.GammaCategorization/CreationalPatterns/5.Singleton/Aula07_SingletonExcercise/Aula_08_SinlgetonExercise2/Main.java
public class Main {
    public static void main(String[] args) {
        // Criando objetos dos órgãos
        POLICIA policia = new POLICIA();
        BOMBEIRO bombeiros = new BOMBEIRO();
        SAMU samu = new SAMU();

        // Comprovando que a CentralDeAlertas é a mesma instância através do hashCode()
        System.out.println("=== COMPROVAÇÃO DE INSTÂNCIA ÚNICA (HASHCODE) ===");
        System.out.println("Hash via Polícia:   " + CentralDeAlertas.getInstancia().hashCode());
        System.out.println("Hash via Bombeiros: " + CentralDeAlertas.getInstancia().hashCode());
        System.out.println("Hash via SAMU:      " + CentralDeAlertas.getInstancia().hashCode());
        System.out.println("===============================================\n");

        // Cada órgão envia um alerta
        policia.enviarAlerta("Ocorrência de roubo na Zona Sul.");
        bombeiros.enviarAlerta("Princípio de incêndio em edificação comercial.");
        samu.enviarAlerta("Acidente de trânsito grave na Rodovia Principal.");

        // Listando a partir de órgãos diferentes para provar o compartilhamento do histórico
        System.out.println("--- Histórico visualizado pela POLÍCIA ---");
        policia.listarAlertas().forEach(System.out::println);

        System.out.println("\n--- Histórico visualizado pelo SAMU ---");
        samu.listarAlertas().forEach(System.out::println);
    }
}